Earthfile

```diff
@@ -39,241 +63,273 @@ setup:
+### Sources Stage
+# This stage is responsible for fetching the ROS 2 workspace sources.
+###############################################################################
 sources:
   FROM +setup
-  COPY +spaceros-artifacts/ros2.repos ros2.repos
-  COPY +spaceros-artifacts/excluded-pkgs.txt excluded-pkgs.txt
-  COPY +spaceros-artifacts/exact.repos exact.repos
-  RUN mkdir src
-  RUN vcs import --shallow --retry 3 --input exact.repos src
-  SAVE ARTIFACT src AS LOCAL src
 
-vcs-exact:
-  COPY +spaceros-artifacts/${SPACEROS_DIR}/exact.repos exact.repos
-  SAVE ARTIFACT exact.repos AS LOCAL exact.repos
+  RUN apt install -y python3-vcstool \
+    && mkdir src -p \
+    && vcs import --retry 3 src < output.repos \
+    && vcs export --exact src > exact.repos
+
+  # Save artifacts to be  used
+  SAVE ARTIFACT exact.repos
+  SAVE ARTIFACT src
 
+###############################################################################
+### Rosdep Stage
+# This stage is responsible for installing the system dependencies required
+# for the ROS 2 workspace.
+###############################################################################
 rosdep:
-  FROM +sources
-  COPY src src/
+  FROM +pre-installation
+
+  # Rosdep updates
+  RUN apt-get install -y python3-rosdep \
+    && rosdep init \
+    && rosdep update
+
+  # Copy Repos file
+  COPY +sources/src ./src
+  COPY excluded-pkgs.txt excluded-deps.txt ./
 
   # Install system package dependencies using rosdep
-  RUN sudo rosdep init && rosdep update
-  RUN --mount=type=cache,mode=0777,target=/var/cache/apt,sharing=locked,id=cache_apt_cache \
-      --mount=type=cache,mode=0777,target=/var/lib/apt,sharing=locked,id=lib_apt_cache \
-      sudo apt-get update && \
-      rosdep update && \
-      rosdep install -y \
+  RUN rosdep install -y \
         --from-paths src --ignore-src \
+        --simulate \
         --rosdistro ${ROS_DISTRO} \
         # `urdfdom_headers` is cloned from source, however rosdep can't find it.
         # It is because package.xml manifest is missing. See: https://github.com/ros/urdfdom_headers
         # Additionally, IKOS must be excluded as per: https://github.com/space-ros/docker/issues/99
-        --skip-keys "$(tr '\n' ' ' < 'excluded-pkgs.txt') urdfdom_headers ikos"
-  RUN rm excluded-pkgs.txt
+        --skip-keys "$(tr '\n' ' ' < 'excluded-pkgs.txt') urdfdom_headers ikos" > rosdeps.txt
+
+  # Process rosdeps.txt to a shell script
+  RUN touch rosdeps.sh \
+        && echo "#!/bin/bash" > rosdeps.sh \
+        && echo "apt-get update" >> rosdeps.sh \
+        && echo "apt-get install -y \\" >> rosdeps.sh \
+        && grep -v -F -f excluded-deps.txt rosdeps.txt | sed 's/^/  /' >> rosdeps.sh \
+        && chmod +x rosdeps.sh \
+        && bash rosdeps.sh
+
+  # The generated shell script is used by the prepare-image stage prior to building the image
+  # saving build time by not having to install dependencies again.
+  SAVE ARTIFACT rosdeps.sh
```

Comment:
Can you elaborate a little more on this and the gains it buys us vs just running the `prepare-image` from `rosdep` instead of `pre-installation`?

Reply:
The stage `pre-installation` is the first version to be prepared and `rosdep` and `prepare-image` is derived.

The [line](https://github.com/franklinselva/space-ros/blob/23538d13f44a22becdf5acc6275f20d9dd604041/Earthfile#L187) in `rosdep` stage will wait for the `sources` stage to complete. For the `prepare-image`, Instead of waiting for the `rosdep` stage which waits for `sources` from the start, we are spinning some instructions to save a little build time.